### DISCUSSION TITLE
How to import additional configuration in programs.nixvim

### DISCUSSION THREAD

**Discussion by mairs8** [nix-community/nixvim#2444] [General]
how can i simply import additional configuration into programs.nixvim? i am getting `error: The option home-manager.users.username.programs.nixvim.import does not exist.`

```nix
programs.nixvim = {
enable = true;

import = [
  ./keymaps.nix
  ./luasnip.nix
  ./lsp.nix
  ./dashboard.nix
  ./trouble.nix
];
```

**Comment by MattSturgeon** [comment]
You should use `imports` instead of `import` (with an `s`).

Additionally, when you have a module that contains `options` and/or `imports` it's often a good idea (though not always required) to nest your config definitions (in this case `enable = true`) under `config`:

```nix
programs.nixvim = {

  import = [
    ./keymaps.nix
    ./luasnip.nix
    ./lsp.nix
    ./dashboard.nix
    ./trouble.nix
  ];

  config = {
    enable = true;
  };

}
```

That should work, but if it doesn't you can force the module system to treat it as a "proper" module by using a module function:

<details><summary>Module function</summary>
<p>


```nix
programs.nixvim = 
  {
    # Optional args:
    config,
    options,
    lib,
    pkgs,
    ...
  }:
  {
    import = [
      ./keymaps.nix
      ./luasnip.nix
      ./lsp.nix
      ./dashboard.nix
      ./trouble.nix
    ];

    config = {
      enable = true;
    };
  }
```

</p>
</details>

**Reply by MattSturgeon** [reply]
Also, note that the imported modules should not contain the `programs.nixvim` prefix.

That prefix is for use in home-manager/nixos/etc modules, and when you use `programs.nixvim.imports` you are instead importing "nixvim modules".

**Reply by mairs8** [reply]
thanks, in the end i used the second option with a more succinct notation below.

```
programs.nixvim = _: {
    imports = [.....]
```

why is it considered good practice to wrap options in `config`?

**Reply by MattSturgeon** [reply]
> thanks, in the end i used the second option with a more succinct notation below.
> 
> ```
> programs.nixvim = _: {
>     imports = [.....]
> ```

Awesome, glad you found something that works!

I must admit, I didn't think using a function-module would be necessary, but I haven't tested it out... I assume you tried both approaches and the plain attrset didn't work?

> why is it considered good practice to wrap options in `config`?

I didn't mean to speak for others; I don't know if the community at large would agree with me 😅

On reflection, I'm not even sure I consistently follow my own suggestion!

But to clarify anyway:

Personally I think it can help distinguish at a glance which parts of the module are configuration definitions and which parts are options/imports.

If you declare any options in your module, then the module system _requires_ you explicitly put configuration in `config`.

If you just have imports, but no options, then you don't _need_ to explicitly nest your config. You may still want to, depending on what you personally find more readable.
